pkg/daemon/daemon_test.go

```diff
@@ -0,0 +1,390 @@
+
+	fakemcclientset "github.com/openshift/machine-config-operator/pkg/generated/clientset/versioned/fake"
+
+	fakek8s "k8s.io/client-go/kubernetes/fake"
+
+	// hack: this import is needed to access sysBuPci unexported variable
+	_ "unsafe"
+)
+
+//go:linkname sysBusPci github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/utils.sysBusPci
+var sysBusPci string
+
+var FakeSupportedNicIDsCM corev1.ConfigMap = corev1.ConfigMap{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      sriovnetworkv1.SupportedNicIDConfigmap,
+		Namespace: namespace,
+	},
+	Data: map[string]string{},
+}
+
+var SriovDevicePluginPod corev1.Pod = corev1.Pod{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      "sriov-device-plugin-xxxx",
+		Namespace: namespace,
+		Labels: map[string]string{
+			"app": "sriov-device-plugin",
+		},
+	},
+	Spec: corev1.PodSpec{
+		NodeName: "test-node",
+	},
+}
+
+func init() {
+	// Increase verbosity to help debugging failures
+	flag.Set("logtostderr", "true")
+	flag.Set("stderrthreshold", "WARNING")
+	flag.Set("v", "2")
+}
+
+func TestConfigDaemon(t *testing.T) {
+	if os.Getuid() != 0 {
+		t.Skip("Config Daemon tests requires root privileges as they leverage netlink functionalities.")
+	}
+
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Config Daemon Suite")
+}
+
+var _ = Describe("Config Daemon", func() {
+	var stopCh chan struct{}
+	var syncCh chan struct{}
+	var exitCh chan error
+	var refreshCh chan Message
+
+	var cleanFakeFs func()
+
+	var sut *Daemon
+
+	var linkDummy = netlink.Dummy{
+		LinkAttrs: netlink.LinkAttrs{
+			Name: "dummysriov0",
+		},
+	}
+
+	BeforeEach(func() {
+
+		stopCh = make(chan struct{})
+		refreshCh = make(chan Message, 1024)
+		exitCh = make(chan error)
+		syncCh = make(chan struct{}, 64)
+
+		// Fill syncCh with values so daemon doesn't wait for a writer
+		for i := 0; i < 64; i++ {
+			syncCh <- struct{}{}
+		}
+
+		// Create virtual filesystem for Daemon
+		fakeFs := &FakeFilesystem{
+			Dirs: []string{
+				"bindata/scripts",
+				"host/sys/bus/pci/devices/0000:86:00.0",
+			},
+			Symlinks: map[string]string{},
+			Files: map[string][]byte{
+				"/bindata/scripts/enable-rdma.sh":                     []byte(""),
+				"/bindata/scripts/load-kmod.sh":                       []byte(""),
+				"/host/sys/bus/pci/devices/0000:86:00.0/sriov_numvfs": []byte(""),
+			},
+		}
+
+		var err error
+		cleanFakeFs, err = fakeFs.Use()
+		Expect(err).ToNot(HaveOccurred())
+
+		kubeClient := fakek8s.NewSimpleClientset(&FakeSupportedNicIDsCM, &SriovDevicePluginPod)
+		kubeClient.Resources = []*metav1.APIResourceList{{GroupVersion: "v1"}}
+		client := fakesnclientset.NewSimpleClientset()
+		mcClient := fakemcclientset.NewSimpleClientset()
+
+		sut = New("test-node",
+			client,
+			kubeClient,
+			mcClient,
+			exitCh,
+			stopCh,
+			syncCh,
+			refreshCh,
+			utils.Baremetal,
+		)
+
+		p, _ := generic.NewGenericPlugin()
+		sut.enabledPlugins = map[string]plugin.VendorPlugin{generic.PluginName: p}
```

Comment:
we should provide a mock interface for generic plugin and not invoke it here IMO.

UT for generic plugin should be added separately

Reply:
I agree. I can work on a version of these tests with all plugins mocked out. 

Some testing hacks would probably go away we will also get less meaningful tests: the current test version has SriovNetworkNodeState as input and filesystem values/iplinks as output.

I'll use a different PR, so we can see and compare both strategies better. WDYT?